backends/p4fmt/p4fmt.cpp

```diff
@@ -1,15 +1,46 @@
 #include "backends/p4fmt/p4fmt.h"
 
+#include "backends/p4fmt/attach.h"
+#include "backends/p4fmt/p4formatter.h"
 #include "frontends/common/parseInput.h"
 #include "frontends/common/parser_options.h"
+#include "frontends/parsers/parserDriver.h"
 #include "ir/ir.h"
 #include "lib/compile_context.h"
 #include "lib/error.h"
+#include "lib/source_file.h"
 #include "options.h"
-#include "p4formatter.h"
 
 namespace P4::P4Fmt {
 
+std::optional<std::pair<const IR::P4Program *, const Util::InputSources *>> parseProgram(
+    const ParserOptions &options) {
+    auto *file = fopen(options.file.c_str(), "r");
+    if (file == nullptr) {
+        ::P4::error(ErrorType::ERR_NOT_FOUND, "%1%: No such file or directory.", options.file);
```

Comment:
why "directory" Are we expect to process a directory here?

Reply:
no. resolved.